pwndbg/storage.py

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+Provides persistent volatile and non-volatile storage of the results of
+a function call.  Volatile storage uses the temporary directory, and is
+generally cleared each reboot, or as the OS dictates.
+
+Cached results are stored per-architecture, i.e. arm and i386 caches
+are independent.  Cached results are also stored per-argument.
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import functools
+import inspect
+import os
+import pickle
+import shelve
+import tempfile
+import time
+
+import pwndbg.arch
+import pwndbg.config
+
+nonvolatile = pwndbg.config.Parameter('pwndbg-cache-nonvolatile', '~/.pwndbg', 'Non-volatile cache directory for Pwndbg storage')
+volatile = pwndbg.config.Parameter('pwndbg-cache-volatile', tempfile.gettempdir(), 'Volatile cache directory for Pwndbg storage')
+
+def name(function):
+    mod_name = function.__module__
+    func_name = function.__name__
+    return mod_name + '.' + func_name
+
+def path(function, directory=volatile):
+    return os.path.join(str(directory), name(function))
+
+class Cached(object):
+    """Function decorator that persistently caches the results of the function it decorates."""
+    function = None
+    type = volatile
```

Comment:
Right now `Cached` and `Volatile` does the same thing.

To prevent one from randomly choosing `Cached` over `Volatile`, the `Cached` should be an abstract class.

However, at least in Py2 it is not possible to have an ABC without a method or property:
```python
In [3]: class A(object):
   ...:     __metaclass__ = ABCMeta
   ...:     

In [4]: class B(A):
   ...:     pass
   ...: 

In [5]: A()
Out[5]: <__main__.A at 0x7fe4c3aa2610>

In [6]: B()
Out[6]: <__main__.B at 0x7fe4c3aac590>
```

So this can be fixed by moving `type` into an `abstractproperty`:
```python
In [19]: class A(object):
    ...:     __metaclass__ = ABCMeta
    ...:     @abstractproperty
    ...:     def type(self):
    ...:         pass
    ...:     

In [20]: class B(A):
    ...:     @property
    ...:     def type(self):
    ...:         return "This is B"
    ...:     

In [21]: A()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-21-3cd318a12eea> in <module>()
----> 1 A()

TypeError: Can't instantiate abstract class A with abstract methods type

In [22]: B().type
Out[22]: 'This is B'
```